fgpyo/alignment/__init__.py

```diff
@@ -0,0 +1,232 @@
+                sequences is gapped.
+            pad_char: The character to use in the query or target sequence lines when
+                padding is required.
+
+        Returns:
+            Three lines representing the alignment.
+        """
+        query_buffer = []
+        align_buffer = []
+        target_buffer = []
+
+        q_offset = self.query_start - 1
+        t_offset = self.target_start - 1
+
+        for cig in self.cigar.elements:
+            length = cig.length
+
+            if cig.operator == CigarOp.I:
+                for _ in range(length):
+                    query_buffer.append(self.query[q_offset])
+                    align_buffer.append(gap_char)
+                    target_buffer.append(pad_char)
+                    q_offset += 1
+            elif cig.operator == CigarOp.D:
+                for _ in range(length):
+                    query_buffer.append(pad_char)
+                    align_buffer.append(gap_char)
+                    target_buffer.append(self.target[t_offset])
+                    t_offset += 1
+            else:
+                for _ in range(length):
+                    q = self.query[q_offset]
+                    t = self.target[t_offset]
+                    query_buffer.append(q)
+                    target_buffer.append(t)
+
+                    if cig.operator == CigarOp.EQ:
+                        align_buffer.append(match_char)
+                    elif cig.operator == CigarOp.X:
+                        align_buffer.append(mismatch_char)
+                    elif cig.operator == CigarOp.M:
+                        align_buffer.append(match_char if q == t else mismatch_char)
+                    else:
+                        raise ValueError(f"Unsupported cigar operator: {cig.operator}")
+
+                    q_offset += 1
+                    t_offset += 1
+
+        return ["".join(query_buffer), "".join(align_buffer), "".join(target_buffer)]
+
+    def sub_by_query(self, start: int, end: int) -> "Alignment":
+        """Returns a subset of the Alignment representing the region defined by start
+        and end on the query sequence. The returned alignment will contain the entire
+        query and target sequences, but will have adjusted query_start and target_start
+        positions and an updated cigar. The score is set to 0.
+
+        Args:
+            start (int): The 1-based inclusive position of the first base on the query
+                sequence to include.
+            end (int): The 1-based inclusive position of the last base on the query
+                sequence to include.
+
+        Returns:
+            A new Alignment with updated coordinates and cigar.
+        """
+        if not (
+            self.query_start <= start <= self.query_end
+            and self.query_start <= end <= self.query_end
+        ):
+            raise ValueError("start or end is outside of aligned region of target sequence")
```

Comment:
it's more useful to have the conditions separated so you know if it start or end that's invalid.  And why not use `assert`?

```suggestion
        assert self.query_start <= start <= self.query_end, f"start '{start}' is outside of aligned region of query sequence "
        assert self.query_start <= end <= self.query_end, f"end '{end}' is outside of aligned region of query sequence "
```

Also, we carried over a bug from fgbio (it's query, not target): https://github.com/fulcrumgenomics/fgbio/pull/914